core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+	for i, r := range result {
+		metaDatas[i] = r.MetaData
+	}
+	return metaDatas, nil
+}
+
+// RetrieveConsumedObjects returns all the consumed objects originated from this node
+// ESS only API
+func (store *CouchStorage) RetrieveConsumedObjects() ([]common.ConsumedObject, common.SyncServiceError) {
+	return nil, nil
+}
+
+// RetrieveObjectData returns the object data with the specified parameters
+func (store *CouchStorage) RetrieveObjectData(orgID string, objectType string, objectID string) (io.Reader, common.SyncServiceError) {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	attachment, err := store.getAttachment(id)
+	if err != nil {
+		switch err {
+		case notFound:
+			return nil, nil
+		default:
+			return nil, &Error{fmt.Sprintf("Failed to open file to read the data. Error: %s.", err)}
+		}
+	}
+
+	data, err := ioutil.ReadAll(attachment.Content)
+	defer attachment.Content.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes.NewReader(data), nil
+}
+
+// CloseDataReader closes the data reader if necessary
+func (store *CouchStorage) CloseDataReader(dataReader io.Reader) common.SyncServiceError {
+	return nil
+}
+
+// ReadObjectData returns the object data with the specified parameters
+func (store *CouchStorage) ReadObjectData(orgID string, objectType string, objectID string, size int, offset int64) ([]byte, bool, int, common.SyncServiceError) {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	attachment, err := store.getAttachment(id)
+	if err != nil {
+		if err == notFound {
+			return nil, true, 0, notFound
+		}
+		return nil, true, 0, err
+	}
+
+	data, err := ioutil.ReadAll(attachment.Content)
+	defer attachment.Content.Close()
+	if err != nil {
+		return nil, true, 0, err
+	}
+
+	lod := int64(len(data))
+	if lod <= offset {
+		return make([]byte, 0), true, 0, nil
+	}
+	s := int64(size)
+	eof := false
+	if s >= lod-offset {
+		s = lod - offset
+		eof = true
+	}
+	b := make([]byte, s)
+	copy(b, data[offset:])
+
+	return b, eof, int(s), nil
+}
+
+// StoreObjectData stores object's data
+// Return true if the object was found and updated
+// Return false and no error, if the object doesn't exist
+func (store *CouchStorage) StoreObjectData(orgID string, objectType string, objectID string, dataReader io.Reader) (bool, common.SyncServiceError) {
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	result := &couchObject{}
+	if err := store.getOne(id, result); err != nil {
+		switch err {
+		case notFound:
+			return false, nil
+		default:
+			return false, &Error{fmt.Sprintf("Failed to store the data. Error: %s.", err)}
+		}
+	}
+
+	if result.Status == common.NotReadyToSend {
+		store.UpdateObjectStatus(orgID, objectType, objectID, common.ReadyToSend)
+	}
+
+	if result.Status == common.NotReadyToSend || result.Status == common.ReadyToSend {
+		newID := store.getInstanceID()
+		result.MetaData.DataID = newID
+		result.MetaData.InstanceID = newID
+		result.LastUpdate = time.Now()
+	}
```

Comment:
Do you need to update the result in the db after set dataID, instanceID and lastUpdate by calling upsertObject() func?

Reply:
Right, sorry about this! To correct this, instead of calling upsertObject() again, would it make a logical difference if I updated DataID, InstanceID and LastUpdate time after adding the attachment, i.e when I update the ObjectSize (line 1023)?
And if I do that, this _if_ condition would not be needed, right? (line 1004)
```
if result.Status == common.NotReadyToSend || result.Status == common.ReadyToSend {
```

Reply:
Yes, that will make the logic different. So in this function, the Metadata.DataID and Metadata.InstanceID only need to be set to newID when status is NotReadyToSend or ReadyToSend. So I think upsert should happen right after line 1008. See [this](https://github.com/open-horizon/edge-sync-service/blob/dde1bef314097c720c0026c127f4854b1cc82ee7/core/storage/mongoStorage.go#L1000) in mongoStorage